oggm/core/inversion.py

```diff
@@ -126,6 +131,113 @@ def _inversion_simple(a3, a0):
     return (-a0)**cfg.ONE_FIFTH
 
 
+def shape_factor_huss(widths, heights, is_rectangular):
+    """Compute shape factor for inclusion of lateral drag
+    according to Huss and Farinotti (2012)
+
+    Discouraged to use, as this does not make any difference
+    between parabolic and rectangular beds
+
+    Parameters
+    ----------
+    widths: ndarray of floats
+        widths of the sections
+    heights: float or ndarray of floats
+        height of the sections
+    is_rectangular: bool or ndarray of bools
+        determines, whether section has a rectangular or parabolic shape
+
+    Returns
+    -------
+    shape factor (no units)
+    """
+    return widths / (2 * heights + widths)
+
+
+def shape_factor_adhikari(widths, heights, is_rectangular):
+    """Compute shape factor for inclusion of lateral drag according to
+    Adhikari (2012)
+
+    TODO: should we expand this here to also include
+    the factors suggested for sliding?
+
+    Parameters
+    ----------
+    widths: ndarray of floats
+        widths of the sections
+    heights: ndarray of floats
+        heights of the sections
+    is_rectangular: ndarray of bools
+        determines, whether section has a rectangular or parabolic shape
+
+    Returns
+    -------
+    shape factors (no units), ndarray of floats
+    """
+    # TODO: how to handle zeta > 10? at the moment extrapolation
+    # Table 1 from Adhikari (2012)
+    tabular_zetas = np.array([0.5, 1, 2, 3, 4, 5, 10])
+    factors_rectangular = np.array([0.313, 0.558, 0.790, 0.884,
+                                    0.929, 0.954, 0.990])
+    f_rect = interp1d(tabular_zetas, factors_rectangular,
+                      fill_value='extrapolate')
+    factors_parabolic = np.array([0.251, 0.448, 0.653, 0.748,
+                                  0.803, 0.839, 0.917])
+    f_parab = interp1d(tabular_zetas, factors_parabolic,
+                       fill_value='extrapolate')
+
+    # Ensure bool (for masking)
+    is_rectangular = is_rectangular.astype(bool)
+
+    # TODO: could check for division by 0, but at the moment
+    # this is covered by interpolation and clip, resulting in a factor of 1
+    zetas = widths / heights
+
+    shape_factors = np.ones(widths.shape)
+
+    # TODO: higher order interpolation? (e.g. via InterpolatedUnivariateSpline)
+    shape_factors[is_rectangular] = f_rect(zetas[is_rectangular])
+    shape_factors[~is_rectangular] = f_parab(zetas[~is_rectangular])
+
+    np.clip(shape_factors, 0.2, 1., out=shape_factors)
+
+    return shape_factors
+
+
+def _compute_thick(gdir, a0s, a3, flux_a0, shape_factor, _inv_function):
+    """
+    TODO
+    Content of the original inner loop of the mass-conservation inversion.
+    Extracted to avoid code duplication
+    Parameters
+    ----------
+    gdir
+    a0s
+    a3
+    flux_a0
+    shape_factor
+    _inv_function
+
+    Returns
+    -------
+
+    """
+    a0s_shaped = a0s / (shape_factor ** 3)
```

Comment:
you can safely overwrite ``a0s`` here

Reply:
But then I cannot exploit the precomputed a0s in subsequent loops

Reply:
Sorry, my mistake. Did not realize I was inside the function.